addons/binding/org.openhab.binding.ihc/src/main/java/org/openhab/binding/ihc/internal/ws/IhcClient.java

```diff
@@ -0,0 +1,558 @@
+     *
+     */
+    public void startControllerEventListeners() {
+        if (getConnectionState() == ConnectionState.CONNECTED) {
+            logger.debug("Start IHC / ELKO listeners");
+            resourceValueNotificationListener = new IhcResourceValueNotificationListener();
+            resourceValueNotificationListener.start();
+            controllerStateListener = new IhcControllerStateListener();
+            controllerStateListener.start();
+        }
+    }
+
+    /**
+     * Query project information from the controller.
+     *
+     * @return project information.
+     * @throws IhcExecption
+     */
+    public synchronized WSProjectInfo getProjectInfo() throws IhcExecption {
+        return controllerService.getProjectInfo();
+    }
+
+    /**
+     * Query system information from the controller.
+     *
+     * @return system information.
+     * @throws IhcExecption
+     */
+    public synchronized WSSystemInfo getSystemInfo() throws IhcExecption {
+        return configurationService.getSystemInfo();
+    }
+
+    /**
+     * Query detected RF devices from the controller.
+     *
+     * @return List of RF devices.
+     * @throws IhcExecption
+     */
+    public synchronized List<WSRFDevice> getDetectedRFDevices() throws IhcExecption {
+        return airlinkManagementService.getDetectedDeviceList();
+    }
+
+    /**
+     * Query controller current state.
+     *
+     * @return controller's current state.
+     */
+    public WSControllerState getControllerState() throws IhcExecption {
+        return controllerService.getControllerState();
+    }
+
+    /**
+     * Query project number of segments.
+     *
+     * @return number of segments.
+     */
+    public int getProjectNumberOfSegments() throws IhcExecption {
+        return controllerService.getProjectNumberOfSegments();
+    }
+
+    /**
+     * Query project segmentation size.
+     *
+     * @return segmentation size in bytes.
+     */
+    public int getProjectSegmentationSize() throws IhcExecption {
+        return controllerService.getProjectSegmentationSize();
+    }
+
+    /**
+     * Query project segments data.
+     *
+     * @return segments data.
+     */
+    public WSFile getProjectSegment(int index, int major, int minor) throws IhcExecption {
+        return controllerService.getProjectSegment(index, major, minor);
+    }
+
+    /**
+     * Fetch project file from controller.
+     *
+     * @return project file.
+     */
+    public byte[] loadProjectFileFromControllerAsByteArray() throws IhcExecption {
+        try {
+
+            WSProjectInfo projectInfo = getProjectInfo();
+            int numberOfSegments = getProjectNumberOfSegments();
+            int segmentationSize = getProjectSegmentationSize();
+
+            logger.debug("Number of segments: {}", numberOfSegments);
+            logger.debug("Segmentation size: {}", segmentationSize);
+
+            ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
```

Comment:
You need to close stream resources. Use:
`try (ByteArrayOutputStream byteStream = new ByteArrayOutputStream()) {`.
Same for input streams below.